### PR TITLE
Fix slowness of relayer with 3+ chains and low average block time

### PR DESCRIPTION
**Pull request by romac**
Closes: https://github.com/informalsystems/ibc-rs/issues/2008
Closes: #1978 
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: relayer/src/util/recv_multiple.rs [hunk 1]
```diff
@@ -1,21 +1,49 @@
...
-    // Receive the message
-    let result = oper.recv(r).ok()?;
+        // Receive the message
+        let result = oper.recv(r).ok()?;
+
+        Some((k, result))
+    }
 
-    Some((k, result))
+    pub fn try_recv_multiple(&mut self) -> Option<(&K, T)> {
```
Comment (adizere) [inline relayer/src/util/recv_multiple.rs:36]: Should we still keep this?

Comment (romac) [reply relayer/src/util/recv_multiple.rs:36]: No I guess we can remove it.

**Comment by ancazamfir** [comment]
Should we try also with https://github.com/informalsystems/ibc-rs/pull/1977?

**Review by adizere — APPROVED** [review]
> > Should we try also with #1977?
> 
> @soareschen is on it :)


Great find Romain, amazing work team! Approving preemptively here.

**Comment by soareschen** [comment]
I have verified that #1977 has been fixed with this PR.

I have also put a fix to correct the use of `spawn_background_task`. In short, the step runner given to `spawn_background_task` is not supposed to block forever. This is because when the `TaskHandle` is dropped, we want to send the shutdown signal to the task runner for it to exit the loop and terminate the task cleanly. If the step runner is blocking forever, then the shutdown signal is never handled, and so the thread join handle never return and cause a deadlock.

**Pull request merged**